client/server/bundler/external-modules.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * External Dependencies
+ */
+const fs = require( 'fs' );
+const path = require( 'path' );
+const { builtinModules } = require( 'module' );
+
+function getModule( request ) {
+	const parts = request.split( '/' );
+	if ( parts[ 0 ].startsWith( '@' ) ) {
+		return parts[ 0 ] + '/' + parts[ 1 ];
+	}
+	return parts[ 0 ];
+}
+
+module.exports = class ExternalModulesWriter {
+	constructor( options ) {
+		this.options = {
+			path: './build',
+			filename: 'modules.json',
+			...options,
+		};
+	}
+
+	apply( compiler ) {
+		compiler.hooks.afterEmit.tapAsync( 'ExternalModulesWriter', ( compilation ) => {
+			const externalModules = new Set();
+
+			for ( const module of compilation.modules ) {
+				if ( ! module.external ) {
+					continue;
+				}
+
+				const requestModule = getModule( module.userRequest );
+
+				// native Node.js module, not in node_modules
+				if ( builtinModules.includes( requestModule ) ) {
+					continue;
+				}
+
+				// loading local file by relative path, not in node_modules
+				if ( requestModule.startsWith( './' ) || requestModule.startsWith( '../' ) ) {
+					continue;
+				}
+
+				externalModules.add( requestModule );
+			}
+
+			fs.writeFileSync(
```

Comment:
Based on [this example](https://webpack.js.org/contribute/writing-a-plugin/#example) from webpack's doc, I think it may be better to use `compilation.assets` to "generate" the file and let webpack write it.

Reply:
I'm not sure if `compilation.assets` is a good fit, because the file is not really a build asset. It's not needed at runtime, and it doesn't make sense to count it towards the total bundle size, which including it as an asset would do.

But when I tried, it worked 🙂 And while debugging it, I discovered webpack's own LibManifestPlugin that uses another method to write files -- webpack internal `compiler.outputFileSystem` object. So I modified the plugin to use that, and get rid of dependencies on `path`, `fs` and `mkdirp`.

`compilation.assets` could be great for the `AssetsWriter` plugin, which produces something that the server really needs at runtime. I found some prior art in the [`react-loadable` library](https://github.com/jamiebuilds/react-loadable/pull/132/files). That library has a plugin that writes information the SSR engine needs at runtime.